example_allocators/bump_allocator/bump_alloc.c

```diff
@@ -0,0 +1,67 @@
+/**********************************
+ * bump_alloc.c
+ * Jeremy.Singer@glasgow.ac.uk
+ *
+ * This is a simple bump-pointer allocator.
+ * It mmaps a large buffer of SIZE bytes,
+ * then allocates this space in word-sized
+ * chunks to client code (in main fn).
+ *
+ * Initial simple memory allocator test.
+ * Explore capability narrowing operations
+ * and intrinsics for bound reporting.
+ */
+
+#include <cheriintrin.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+#include "bump_alloc.h"
+
+int count = 0;		 /* number of bytes allocated so far*/
+int max = 0;		 /* upper limit for count */
+char *buffer = NULL; /* the allocation buffer */
+
+void init_alloc(int size_in_bytes)
+{
+	/* request memory for our allocation buffer */
+	char *res = mmap(NULL, size_in_bytes, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+
+	if (res == MAP_FAILED)
+	{
+		perror("error in initial mem allocation");
+		exit(-1);
+	}
+
+	/* NB mmap min bounds for capability is 1 page (4K) */
+	// if (DEBUG_PRINTF)
+	//	inspect_pointer(res);
+
+	buffer = res;
+	max = size_in_bytes;
+	return;
+}
+
+/*
+ * allocate len bytes with bump pointer allocator
+ * this is our simplistic `malloc` function
+ */
+char *bump_alloc(int len)
+{
+	char *chunk = 0;
+	if (count + len <= max)
+	{
+		/* bump the pointer */
+		chunk = buffer + count;
+
+		/* restrict capability range before returning ptr */
+		chunk = cheri_bounds_set_exact(chunk, len);
```

Comment:
Did you give any thought to how you might give the user _at least_ as much memory as they requested? Aborting with awkward lengths is fine for an example, but not ideal for a real allocator. Simply using `cheri_bounds_set` (not `_exact`) isn't the answer because (if I understand correctly) that might adjust the base downwards, into already-allocated memory.

Reply:
Oh, the CHERI C/C++ Programming Guide has a section (7.5) on this, with an example implementation.

Reply:
thanks @jacobbramley I have added this fixup code now for alignment ... please could you review?